src/PetriEngine/Colored/Reduction/RedRulePreAgglomeration.cpp

```diff
@@ -170,60 +181,170 @@ namespace PetriEngine::Colored::Reduction {
+                        } else {
+                            Colored::VariableVisitor::get_variables(*consumer.guard, consVars);
+                        }
+
+                    }
+                    for (auto& arc : consumer.input_arcs){
+                        if (!consHangingGuardVarRisk || arc.place != consArc->place){
+                            Colored::VariableVisitor::get_variables(*arc.expr, consVars);
+                        }
+                    }
+                    for (auto& arc : consumer.output_arcs){
+                        if (!consHangingGuardVarRisk || arc.place != consArc->place){
+                            Colored::VariableVisitor::get_variables(*arc.expr, consVars);
                         }
-                        if (!ok) continue;
                     }
 
-                    uint32_t w = red.getInArc(pid, consumer)->expr->weight();
+                    if (consHangingGuardVarRisk) {
+                        for (auto& var : consArcVars){
+                            if (consVars.find(var) == consVars.end()){
+                                if (consGuardVars.find(var) != consGuardVars.end()){
+                                    // Both a producer and a consumer have a hanging guard variable, at which point it is an actual problem.
+                                    ok = false;
+                                    break;
+                                } else {
+                                    //
+                                    consVars.insert(var);
+                                }
+                            }
+                        }
+                    }
+
+                    if (!ok) continue;
 
                     // Update
                     for (const auto& prod : originalProducers){
                         const Transition& producer = red.transitions()[prod];
                         const CArcIter proArc = red.getOutArc(producer, pid);
+                        std::set<const Variable*> pairVars;
+                        std::unordered_map<uint32_t, std::vector<const Colored::Variable*>> tuples;
+                        // Valid because there is only allowed to be at most 1 variable on each.
+                        Colored::VariableVisitor::get_variables(*proArc->expr, pairVars, tuples);
+                        Colored::VariableVisitor::get_variables(*consArc->expr, pairVars, tuples);
+
+                        std::vector<Variable*> centerVariables;
+                        std::unordered_map<std::string, const Variable*> varReplacementMap;
+                        for (auto& tuple : tuples){
+                            for (uint32_t tupleIndex = 0; tupleIndex < tuple.second.size(); tupleIndex++){
+                                if (centerVariables.size() <= tupleIndex){
+                                    centerVariables.emplace_back(new Variable{*producer.name + *consumer.name + tuple.second[tupleIndex]->name, tuple.second[tupleIndex]->colorType});
```

Comment:
All variables must be in `cpnBuilder::variables` otherwise the CPN cannot be exported with the PNMLWriter